### ISSUE TITLE
Attempt 1 failed. KeyError: 'contentProperties'

### ISSUE THREAD

**Issue by henfri** [breunigs/duplicity-acdcli#10]
Hello,

I am running duplicity 0.7.06. 
I have downloaded:
`wget -O /usr/lib/python2.7/dist-packages/duplicity/backends/amazonclouddrivebackend.py  https://raw.githubusercontent.com/breunigs/duplicity-acdcli/master/amazonclouddrivebackend.py`
Unfortunately, when running
`duplicity full /mnt2/source/ acd://mnt2/amazon/duplicity/`
I get:

> Local and Remote metadata are synchronized, no sync needed.
> Last full backup date: none
> GnuPG passphrase for decryption:
> Retype passphrase for decryption to confirm:
> Attempt 1 failed. KeyError: 'contentProperties'
> Attempt 2 failed. KeyError: 'contentProperties'
> Attempt 3 failed. KeyError: 'contentProperties'
> Attempt 4 failed. KeyError: 'contentProperties'
> Attempt 5 failed. KeyError: 'contentProperties'
> Attempt 1 failed. KeyError: 'contentProperties'
> 

/mnt2/amazon is my amazon cloud drive:
ACDFuse on /media/a8af3832-48c7-4568-861f-e80380dd7e0b/amazon type fuse.ACDFuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0)

(mnt2 is a symlink to /media/a8af3832-48c7-4568-861f-e80380dd7e0b/)

Do I understand well, that the pgp key can my selected by myself and is not in any way related to the amazon drive?

Thanks for your work on this!

Greetings,
Hendrik

**Comment by breunigs** [comment]
Yes, the GPG key and/or passphrase is not related to Amazon Drive. In fact, they definitely should not be the same, or the encryption is probably useless.

In any case, while this backend requires acd_cli to be setup, mounting using `acd_cli mount /where/ever` is not required. As such if you specify `acd://mnt2/amazon/duplicity/` you would see a `mnt2/amazon/duplicity` like structure even when browsing your Amazon Drive through their website. Put another way, if you specify on duplicity `acd://mnt2/amazon/duplicity/`, your local path will be this:

```
/mnt2/amazon/mnt2/amazon/duplicity/
       - or -
/media/a8af3832-48c7-4568-861f-e80380dd7e0b/mnt2/amazon/duplicity/
```

I guess you want to do something like this:

```
# First, create the folder you want to backup to on Amazon Drive. You can do this through their website, using your ACD mount or directly with acd_cli like shown:
acd_cli mkdir --parents duplicity

# then, backup to that folder:
duplicity full /mnt2/source/ acd:///duplicity
```

**Comment by breunigs** [comment]
If above does not help, are you sure you setup acd_cli correctly for the same user you run duplicity with? If you are unsure, try something like `sudo -u SOME_USER_HERE acd_cli ls` and if it does not list your top level files/folders on Amazon Drive, then this is probably the cause.

**Comment by henfri** [comment]
Hello,

thanks a lot. That was not obvious to me.
In fact, I got a bit further now.


`root@homeserver:~acd_cli mkdir --parents duplicity1
root@homeserver:~#  duplicity full /mnt2/RockOn_Configs/ acd://duplicity1/
Local and Remote metadata are synchronized, no sync needed.
Warning, found incomplete backup sets, probably left from aborted session
Last full backup date: none
GnuPG passphrase for decryption:
Retype passphrase for decryption to confirm:
Attempt 1 failed. KeyError: 'contentProperties'

Attempt 2 failed. KeyError: 'contentProperties'
Attempt 3 failed. KeyError: 'contentProperties'
^Croot@homeserver:~# acd_cli ls
[A3qiHUfYSb2h2vijW55KUw] [A] Bilder/
[SSRISvCqRVacEt3edzgMGw] [A] Dokumente/
[8RcljmiQRIKzRgxYC2biqA] [A] Videos/
[iwLCnY26RrKi4A5gqDgsGw] [A] duplicity/
[ypGeYOqOSvW9lz-01PYBgQ] [A] duplicity1/
[WUNF7-ShTyijTURwYKChVw] [A] duplicity-full.20160728T182810Z.vol1.difftar.gpg
[Xd2_78NfTPex-9JuiD85lA] [A] duplicity-full.20160728T182810Z.vol2.difftar.gpg
[UiOx3XVSSPGqHwsQ5_T4OA] [A] duplicity-full.20160728T183605Z.vol1.difftar.gpg
[tq3BjBW_QMuxaP79mjoQCA] [A] duplicity-full.20160728T185232Z.vol1.difftar.gpg
[nq2S3YJiTHO8ehavEOCKHA] [A] duplicity-full.20160728T190436Z.vol1.difftar.gpg
[RYdWro5kRY6W85GDxzHHJQ] [A] duplicity-full.20160728T190800Z.vol1.difftar.gpg
[jCqGzdRUTiK5NxRN5GHnsA] [A] duplicity-full.20160728T190852Z.vol1.difftar.gpg
`
I am running both as root.

Do you have any other idea?

Greetings,
Hendrik

**Comment by breunigs** [comment]
Hm, not from the top of my head. You can try to manually clean up all of the test files you have created and start fresh, to see if the problem persists.

If you do, please run duplicity with verbose output, like so:
```
duplicity -v9 full /mnt2/RockOn_Configs/ acd://duplicity1/
```
and post the result here. This output includes your (presumably) private filenames, so maybe run this just with test data.

**Comment by henfri** [comment]
Hello,

thanks for your help.
As far as I read the logs, the backup does succeed, doesn't it?
http://paste.debian.net/785849/

Greetings,
Hendrik

**Comment by breunigs** [comment]
I'm still not quite sure what happens here. Everything looks okay, it seems the upload works -- but reading the metadata right after does not anymore.

Which version of acd_cli do you use: `acd_cli v`?

Does the issue go away when you delete acd_cli's local cache: `rm ~root/.cache/acd_cli/nodes.* && acd_cli sync`?

Is the ACDFuse mounted with the root user as well? If so, you *must* unmount it before running duplicity, as concurrent access (more precisely: concurrent syncing) is not supported by acd_cli and will lead to weird errors. Alternatively, run either the mount or duplicity with a different user, so that acd_cli will use a different database when syncing.

If the problem persists after this, please unmount all acd_cli and do roughly the following:
```
rm ~root/.cache/acd_cli/nodes.*
acd_cli sync
rm ~root/.cache/acd_cli/*.log
duplicity -v9 full /mnt2/RockOn_Configs/ acd://duplicity1/
```
and upload acd_cli's log files.

**Comment by breunigs** [comment]
If you want to ensure the backup worked, you can always try to restore files from it to some other directory. However, it seems rather weird that it retries to read the data for so long, and then just gives up without any complaint?

**Comment by henfri** [comment]
Hey,

thanks for your help.
Indeed it was mounted. 
Unmounting did not help.
`rm ~root/.cache/acd_cli/nodes.* && acd_cli sync`
did not help aswell.

The version:
```
root@homeserver:~# acd_cli v
acd_cli 0.3.1, api 0.8.7
```

But this:
`rm ~root/.cache/acd_cli/nodes.*
acd_cli sync
rm ~root/.cache/acd_cli/*.log
duplicity -v9 full /mnt2/RockOn_Configs/ acd://duplicity1/`

did make a difference.  Not a success, but an other error though:
`Main action: full
================================================================================
duplicity 0.7.06 (December 07, 2015)
Args: /usr/bin/duplicity -v9 full /mnt2/RockOn_Configs/ acd://duplicity77/
Linux homeserver 4.6.0-0.bpo.1-amd64 #1 SMP Debian 4.6.3-1~bpo8+1 (2016-07-13) x86_64
/usr/bin/python2 2.7.9 (default, Mar  1 2015, 12:57:24)
[GCC 4.9.2]
================================================================================
Using temporary directory /tmp/duplicity-tm98S7-tempdir
Registering (mkstemp) temporary file /tmp/duplicity-tm98S7-tempdir/mkstemp-fCFLRB-1
Temp has 9664507904 available, backup will use approx 34078720.
Reading results of 'acd_cli ls '//duplicity77/''
Synchronizing remote metadata to local cache...
PASSPHRASE variable not set, asking user.
Copying duplicity-full-signatures.20160728T205130Z.sigtar.gpg to local cache.
Registering (mktemp) temporary file /tmp/duplicity-tm98S7-tempdir/mktemp-1HdziA-2
Reading results of 'acd_cli download '//duplicity77/duplicity-full-signatures.20160728T205130Z.sigtar.gpg' '/tmp/duplicity-tm98S7-tempdir''
Backtrace of previous error: Traceback (innermost last):
  File "/usr/lib/python2.7/dist-packages/duplicity/backend.py", line 365, in inner_retry
    return fn(self, *args)
  File "/usr/lib/python2.7/dist-packages/duplicity/backend.py", line 563, in get
    self.backend._get(remote_filename, local_path)
  File "/usr/lib/python2.7/dist-packages/duplicity/backends/amazonclouddrivebackend.py", line 103, in _get
    os.rename(os.path.join(local_dir, remote_filename), local_path.name)
 OSError: [Errno 2] No such file or directory

Releasing lockfile /root/.cache/duplicity/d4d7054f1ab3a69f7197646062b69ab4/lockfile.lock
Removing still remembered temporary file /tmp/duplicity-tm98S7-tempdir/mktemp-1HdziA-2
Removing still remembered temporary file /tmp/duplicity-tm98S7-tempdir/mkstemp-fCFLRB-1
Cleanup of temporary directory /tmp/duplicity-tm98S7-tempdir failed - this is probably a bug.
`

Where do I find acd_cli's log files?
I could not find a reference in the docs (http://acd-cli.readthedocs.io/en/latest/search.html?q=log&check_keywords=yes&area=default)

Regards,
Hendrik

**Comment by henfri** [comment]
Never mind, I found the logs in  ./.cache/acd_cli/acd_cli.log
http://paste.debian.net/785860/

Greetings,
Hendrik

**Comment by breunigs** [comment]
Please update acd_cli, it should output or higher: `acd_cli 0.3.1, api 0.8.9`

The logs seem right, nothing out of the ordinary that I can see. Although there's still some process running that is using `root`'s acd_cli instance to download a file called `s?e? -  Exil.mkv`. Are you sure you unmounted acd_cli elsewhere and stopped scripts running in the background? `ps -A f -o pid,user,args` can help you figure out what is still running on your machine.

The error duplicity reported is that it cannot find the file it tried to download just now. I don't understand how this happens, since there is no mention of any failure in acd_cli's log.

What is the output of this:

```
acd_cli download '//duplicity77/duplicity-full-signatures.20160728T205130Z.sigtar.gpg' '/tmp/'
```

and after the download

```
ls -alh '/tmp/duplicity-full-signatures.20160728T205130Z.sigtar.gpg'
```

**Comment by henfri** [comment]
Hello,

I think I got it.
I did update acd_cli_
pip3 install --upgrade git+https://github.com/yadayada/acd_cli.git
This broke pip:
ImportError: cannot import name 'IncompleteRead'
This was fixed by:
apt-get remove python3-pip
easy_install3 pip

When then running acd_cli I had the problem, that the db of acd_cli was locked.
I fixed that by killing the process that had mounted the amazon drive (fuse mount).

After this, duplicity seems to run.

I'll monitor and further report.
Thanks for the time being!

Greetings,
Hendrik
